v2/writer.go

```diff
@@ -0,0 +1,125 @@
+package car
+
+import (
+	"bytes"
+	"context"
+	"github.com/ipfs/go-cid"
+	format "github.com/ipfs/go-ipld-format"
+	car_v1 "github.com/ipld/go-car"
+	"github.com/willscott/carbs"
+	"io"
+)
+
+type (
+	// Padding represents the number of padding bytes.
+	Padding uint64
+	// Writer writes CAR v2 into a give io.Writer.
+	Writer struct {
+		Walk         car_v1.WalkFunc
+		IndexCodec   carbs.IndexCodec
+		NodeGetter   format.NodeGetter
+		CarV1Padding Padding
+		IndexPadding Padding
+
+		ctx   context.Context
+		roots []cid.Cid
+	}
+)
+
+// WriteTo writes this padding to the given writer as default value bytes.
+func (p Padding) WriteTo(w io.Writer) (n int64, err error) {
+	paddingBytes := make([]byte, p)
+	written, err := w.Write(paddingBytes)
+	n = int64(written)
+	return
+}
+
+// NewWriter instantiates a new CAR v2 writer.
+// The writer instantiated uses `carbs.IndexSorted` as the index codec,
+// and `car_v1.DefaultWalkFunc` as the default walk function.
+func NewWriter(ctx context.Context, ng format.NodeGetter, roots []cid.Cid) *Writer {
+	return &Writer{
+		Walk:       car_v1.DefaultWalkFunc,
+		IndexCodec: carbs.IndexSorted,
+		NodeGetter: ng,
+		ctx:        ctx,
```

Comment:
not too excited about the writer taking a context, but I get that WriterTo taking a context itself would break the io interface.

I guess our use case is weird, because the writing incurs IPLD walking, which should use a context. Most WriterTos don't need a context to write themselves.

I assume we're fine with this given that a Writer is short-lived. Should we make WriteTo error if it's already been called? Using a single Writer to write many times feels wrong, because that would mean reusing the same context, which is almost always not what you want. It also seems wrong to want to marshal exactly the same CAR multiple times in a row: people should marshal it once, and reuse the bytes.

Reply:
Another option is to give up io.WriterTo; the user, instead of writing:

    n, err := carv2.NewWriter(ctx, ng, roots).WriteTo(w)

would write something like:

    n, err := carv2.Marshal(ctx, w, ng, roots)

I lean slightly in favor of the second form, because it avoids the anti-pattern of holding a context in a struct field, and avoids possible misuses of the API like holding onto a Writer after the marshaling has completed.

The only possible downside to my mind is that we can't add more parameters or options in the future. For that reason, we could add a `...MarshalOption` as a last parameter, so in the future we could expose marshal option funcs for any possible options we might need.

Reply:
Sounds good thanks @mvdan I am with you on ditching the `io.Writer` I think it's only a matter of time for us to have to break it at some point.

Reply:
@mvdan if you don't mind I address this in a sparate PR?